### ISSUE TITLE
ProductProposalsAccessForbiddenException - please, activate my account

### ISSUE THREAD

**Issue by fazrik** [allegro/allegro-api#8199]
Hi,
I have the problem with ProductProposalsAccessForbiddenException.
My Sandbox ClientID: 106804232 ([marek@webyslavik.cz](mailto:marek@webyslavik.cz)).
Could you fully activate my account, please?
Thanks, Marek

**Comment by MaciejFrackowiak** [comment]
If you want to list an offer through the API, you can switch your account to the company type. You can do it in the [Account details](https://allegro.pl.allegrosandbox.pl/moje-allegro/moje-konto/dane-konta) tab.

**Comment by fazrik** [comment]
Thanks for an advice, but if you mean Change to [Business Account](https://allegro.pl.allegrosandbox.pl/biznes/konwersja-konta), I have to fill Polish NIP (Tax Identification Number) in the next step and I do not know one. When I try [Activate sales](https://allegro.pl.allegrosandbox.pl/lokalnie/c2c/sales-activation), [webpage c2c-payu-simulator.dh-test.qxlint](http://c2c-payu-simulator.dh-test.qxlint/simulator/orders/9b7f95e6-ec7f-4e02-a9cf-7db5e291fee7) does not work.

**Comment by MaciejFrackowiak** [comment]
On the Sandbox you should use fiction data - there are plenty of simulators available in the network, just use one of them. Also, you do not have to choose Polish data - you can provide Czech instead.

**Comment by fazrik** [comment]
I have switched my account to Business, but ProductProposalsAccessForbiddenException still appears.

**Comment by MaciejFrackowiak** [comment]
Unfortunately, you chose the specific company type "foundation", which is not a typical business company. Please, one more time switch to another e.g. "sole trader", and then I will be able to activate your account.

**Comment by fazrik** [comment]
I'm not able to switch account from Business in Account / Settings /
[Account Details](https://allegro.pl.allegrosandbox.pl/moje-allegro/moje-konto/dane-konta), because Business Account phrase is not a link:
ACCOUNT TYPE
Business Account
SALES
Inactive
You cannot list offers
[Activate sales](https://allegro.pl.allegrosandbox.pl/rejestracja/konto-firmowe/ustaw-wyplaty)

**Comment by MaciejFrackowiak** [comment]
OK, now you should make a bank transfer and send any files as company documents. You can do it via the [Account details](https://allegro.pl.allegrosandbox.pl/moje-allegro/moje-konto/dane-konta) tab.

**Comment by fazrik** [comment]
Should I use real account, will I really pay money in Sandbox environment?

**Comment by MaciejFrackowiak** [comment]
No, you should use fiction data as I wrote. Also, on the Sandbox there is a payment simulator, so you do not have to make a real bank transfer.

**Comment by fazrik** [comment]
I've received confirmation mail 

All payouts are withheld until the new account has been positively verified.
CONFIRM THE NEW ACCOUNT 
https://t.allegro.pl.allegrosandbox.pl/moje-allegro/sprzedaz/ustawienia-wyplat/potwierdz-nowe-konto?utm_source=notification&utm_medium=payoutAccountChanged&utm_campaign=de0b263b-3cb1-4e5a-b0c4-d14aa9e1027d&tr_n=confirm-new-payout-account&tr_id=106804232-v0&tr_d=allegro.pl&tr_c=payout-settings&tr_s=it1Uc1jg

which is redirected to https://allegro.pl/logowanie...
so I'm not sure that confirmation succeeded.

**Comment by MaciejFrackowiak** [comment]
Your account is active now.

**Comment by fazrik** [comment]
Thank you, now I'm able to list offers.

**Comment by fazrik** [comment]
Hi Maciej, 

I have listed the products and before the operational launch I would like to clarify 
the options for running updates of offers and importing orders via cron.

* what is the recommended way for cron-driven API communication?
* how often is Device confirmed for API requests (Device Application)?
* how often would user have to confirm API requests by logging into Allegro (Web/App with browser access Application)?

Thank you, you can answer to marek@webyslavik.cz, if you want.

Regards, Fazrik

**Comment by MaciejFrackowiak** [comment]
I am glad that you successfully listed an item. We do not have any special recommendations - it depends on your specific way of operation. If any user gives you consent to operate in his name, as long as the token is valid, you can operate.

**Comment by fazrik** [comment]
I have created Device application, I have authorized device via /uzytkownik/bezpieczenstwo/skojarz-aplikacje?code=xxxxx URL and I haven't had to authorize sandbox communication again in any way for at least two weeks as well as when I communicate through the Browser Device. I do not understand, if Device is authorized only once or if it will have to be re-authorized after some time.

**Comment by PrzemyslawLukanowski** [comment]
Hello, the access token you received is valid only for 12 hours, using this token you can perform within this time operations on API resources on behalf of an authorized user. You can check the token details, such as expiration date, by decoding it, for example, [on this page](https://jwt.io/). To avoid the user having to manually perform application authorization twice a day, you can implement the[ refresh token mechanism](https://developer.allegro.pl/tutorials/authentication-and-authorization-m09BlVyo7iY#extending-token-validity), thanks to which the authorization is needed to be done only once, at the beginning.

**Comment by fazrik** [comment]
Thank you. One other thing: I'm not able to use any Authorized user for approving REST API communication, because I'm getting error ShippingRatesNotFoundException. What do I do wrong? It is not acceptable for us to create shipping rates lists for every Authorized user. Can you help me?

**Comment by PrzemyslawLukanowski** [comment]
Please check the information from [our tutorial](https://developer.allegro.pl/tutorials/list-offer-assigned-product-one-request-D7Kj9M71Bu6):
<img width="626" alt="Zrzut ekranu 2024-01-9 o 13 56 50" src="https://github.com/allegro/allegro-api/assets/51473661/7f35be20-c5c4-425b-9f84-f2e5c6cbdbbc">
It seems that the shipping rate with the name "default" was not set on the given account. 
So, you have two options:
- users should have defined "default" shipping rate on their accounts - that way, there will be nothing to change in the JSON structure, cause we will include it automatically;
- you can extend your requests with the shipping rate id or name, [here](https://developer.allegro.pl/tutorials/list-offer-assigned-product-one-request-D7Kj9M71Bu6#shipping-location-and-price-lists) you can find some examples:

name:
```
...
"delivery": {
 "shippingRates": {
     "name": "small dimensions"
 }
},
...
```
or ID:
...
```
"delivery": {
 "shippingRates": {
     "id": "c446793c-33f0-407f-b0ed-1aeec6090a7a"
 }
}
```
...

**Comment by fazrik** [comment]
After adding shippingRates to object, I'm getting ProductProposalsAccessForbiddenException. I'm Authorised user of my client's account with full rights. My account name is Fazrik (marek@webyslavik.cz), client's account name is intimity_cz (info@intimity.cz)

**Comment by PrzemyslawLukanowski** [comment]
Ok, thanks for that info. You are trying to list a new offer being authorized as 'Fazrik' - this account is a regular type (not a company) and to list new offers via API you need to have a company account ([our news](https://developer.allegro.pl/news/from-today-only-users-with-company-account-can-list-and-edit-offers-via-api-zlmZkb8xGiw) from the past associated with this change). When it comes to the client account - 'intimity_cz', this is a company account, but I don't see in our logs any operation on behalf of this user.

**Comment by fazrik** [comment]
I have listed 141 offers on behalf of user intimity_cz today via Intimity_Browser (ID: 115d7f9780c84aee9dfb0fadf366e104). I don't understand how to use Authorized user for approving REST API communicaton.

**Comment by PrzemyslawLukanowski** [comment]
You received a 403 Forbidden status code cause you used an access token that was generated for the 'Fazrik' user, which is a regular account. As you said and I also confirm by checking our logs, the intimity_cz account properly authorized your application, cause we can see the traffic from this account via your application (the last logs are from 12:35). Please make sure you are using the correct token within the REST API communication - you can call [GET /me](https://developer.allegro.pl/documentation#operation/meGET) to check who the token belongs to.

**Comment by fazrik** [comment]
I've generated new token for Fazrik and now I'm getting ProductProposalsAccessForbiddenException again. So it looks like Fazrik (or any other Authorized user) is unusable for REST API commnication. Am I right?

**Comment by PrzemyslawLukanowski** [comment]
Yes, Fazrik is a regular account and to list new offers and manage them via API is necessary to have a company account.

**Comment by fazrik** [comment]
Hi, how can I solve this type of error when the brand "Chung shi" is the correct one for this product 
and "other" certainly is not? Thank you. 

PARAMETER_MISMATCH: The specified product already exists. The given parameter "Brand" with the value "Chung shi (7108_1790571)" does not match the product parameter "other (7108_52)".

**Comment by MaciejFrackowiak** [comment]
You can propose a change in the existing product using [POST /sale/products/{productId}/change-proposals](https://developer.allegro.pl/documentation/#operation/productChangeProposal).

**Comment by fazrik** [comment]
When I try to POST Change proposal, I get NotAllowedException [message] => The HTTP method is not allowed.
There is allegro:api:sale:offers:write permission in scope item of token response:

[access_token] => eyJhbGc...
[token_type] => bearer
[refresh_token] => eyJhbGc...
[expires_in] => 43199
[scope] => allegro:api:orders:read allegro:api:fulfillment:read allegro:api:profile:write allegro:api:sale:offers:write allegro:api:fulfillment:write allegro:api:billing:read allegro:api:campaigns allegro:api:disputes allegro:api:bids allegro:api:shipments:write allegro:api:sale:offers:read allegro:api:orders:write allegro:api:ads allegro:api:payments:write allegro:api:sale:settings:write allegro:api:profile:read allegro:api:ratings allegro:api:sale:settings:read allegro:api:payments:read allegro:api:shipments:read allegro:api:messaging
[allegro_api] => 1
[iss] => https://allegro.pl
[jti] => bae25ecc-238e-4693-8b5b-a061c012277f

**Comment by fazrik** [comment]
Hi, what is the recommended way to prevent the possibility of buying a product that is temporarily out of stock?
I'm not able to send quantity=0 and to expire product looks strange to me.
Thank you
